Compatibility with future PT 1.12

## 🐛 Bug

```
>       distance = x @ y.T
E       RuntimeError: "addmm_impl_cpu_" not implemented for 'Half'
```

### To Reproduce

https://github.com/PyTorchLightning/metrics/runs/6275393755?check_suite_focus=true

#### Code sample

```py
distance = x @ y.T
```

### Expected behavior

<!-- A clear and concise description of what you expected to happen. -->

### Environment

- OS (e.g., Linux): linux
- Python & PyTorch Version (e.g., 1.0): py3.8 & pt1.12

### Additional context

context see https://github.com/pytorch/pytorch/pull/75647
slack: https://pytorch.slack.com/archives/C3PDTEV8E/p1651742487294399

tl;dr: We don't have correct and fast linalg algorithms for half in CPU as none of the backend libraries that we use support them so it's better not to support these via half-baked implementations, and simply make the users cast their inputs if they want to use these methods

> but for GPU it shall be supported, correct? >> Yes

note as well that this just applies to Half on CPU and for linalg functios (think matmul and conv). Pointwise functions on Half on CPU will still be available, and Half on CUDA will still have full support

> it was implemented up till 1.11 but there was no real speed-up, correct?

Not only it was slower, but it was not numerically stable, so it was pretty much a bug (hence the removal without deprecation)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -63,7 +63,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Fixed multi device aggregation in `PearsonCorrCoef` ([#998](https://github.com/PyTorchLightning/metrics/pull/998))
 
 
--
+- Fixed compatibility with future Pytorch 1.12 in `pairwise_cosine_similarity` ([#1011](https://github.com/PyTorchLightning/metrics/pull/1011))
 
 
 ## [0.8.1] - 2022-04-27

--- a/torchmetrics/functional/pairwise/cosine.py
+++ b/torchmetrics/functional/pairwise/cosine.py
@@ -20,6 +20,16 @@ from typing_extensions import Literal
 from torchmetrics.functional.pairwise.helpers import _check_input, _reduce_distance_matrix
 
 
+def _safe_matmul(x: Tensor, y: Tensor) -> Tensor:
+    """Safe calculation of matrix multiplication.
+
+    If input is float16, will cast to float32 for computation and back again.
+    """
+    if x.dtype == torch.float16 or y.dtype == torch.float16:
+        return (x.float() @ y.T.float()).half()
+    return x @ y.T
+
+
 def _pairwise_cosine_similarity_update(
     x: Tensor, y: Optional[Tensor] = None, zero_diagonal: Optional[bool] = None
 ) -> Tensor:
@@ -37,7 +47,7 @@ def _pairwise_cosine_similarity_update(
     norm = torch.norm(y, p=2, dim=1)
     y /= norm.unsqueeze(1)
 
-    distance = x @ y.T
+    distance = _safe_matmul(x, y)
     if zero_diagonal:
         distance.fill_diagonal_(0)
     return distance

